pkg/server/distributor_test.go

```diff
@@ -0,0 +1,475 @@
+			ctx = identity.WithServiceIdentityContext(context.Background(), 1)
+			var header metadata.MD
+			res, err := distributorServer.resourceClient.Search(ctx, &resourcepb.ResourceSearchRequest{
+				Options: &resourcepb.ListOptions{
+					Key: &resourcepb.ResourceKey{
+						Group:     "playlist.grafana.app",
+						Resource:  "aoeuaeou",
+						Namespace: ns,
+					},
+				},
+			}, grpc.Header(&header))
+			require.NoError(t, err)
+
+			// sometimes the querycost is different between the two. Happens randomly and we don't have control over it
+			// as it comes from bleve. Since we are not testing search functionality we hard-set this to 0 to avoid
+			// flaky tests
+			res.QueryCost = 0
+			baselineRes.QueryCost = 0
+
+			require.Equal(t, baselineRes.String(), res.String())
+
+			instance := header.Get("proxied-instance-id")
+			if len(instance) != 1 {
+				t.Fatal("received invalid proxied-instance-id header", instance)
+			}
+
+			instanceResponseCount[instance[0]] += 1
+		}
+
+		for instance, count := range instanceResponseCount {
+			require.GreaterOrEqual(t, count, 1, "instance did not get any traffic: "+instance)
+		}
+	})
+
+	stopErrs := make(map[string]error)
+	stopServers := func(done chan error) {
+		var wg sync.WaitGroup
+		for _, testServer := range testServers {
+			wg.Add(1)
+			go func(s testModuleServer) {
+				defer wg.Done()
+				ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+				defer cancel()
+				if err := testServer.server.Shutdown(ctx, "tests are done"); err != nil {
+					stopErrs[testServer.id] = err
+				}
+			}(testServer)
+		}
+		wg.Wait()
+
+		ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+		defer cancel()
+		if err := distributorServer.server.Shutdown(ctx, "tests are done"); err != nil {
+			stopErrs[distributorServer.id] = err
+		}
+		done <- nil
+	}
+
+	done := make(chan error, 1)
+	go stopServers(done)
+	select {
+	case <-done:
+	case <-time.After(30 * time.Second):
+		t.Fatal("timeout waiting for servers to shutdown")
+	}
+
+	for server, runErr := range runErrs {
+		t.Fatalf("unexpected run error from module server %s: %v", server, runErr)
+	}
+
+	for server, stopErr := range stopErrs {
+		t.Fatalf("unexpected stop error from module server %s: %v", server, stopErr)
+	}
+}
+
+func startAndWaitHealthy(t *testing.T, testServer testModuleServer, runErrs map[string]error) {
+	go func() {
+		// this next line is to avoid double registration, as both InitializeDocumentBuilders as well as ProvideUnifiedStorageGrpcService
+		// are hard-coded to use prometheus.DefaultRegisterer
+		// the alternative would be to get the registry from wire, in which case the tests would receive a new
+		// registry automatically, but that _may_ change metric names
+		// We can remove this once that's fixed
+		prometheus.DefaultRegisterer = prometheus.NewRegistry()
+		if err := testServer.server.Run(); err != nil && !errors.Is(err, context.Canceled) {
+			runErrs[testServer.id] = err
```

Comment:
We're accessing `runErrs` from two goroutines here without synchronization.

Instead of collecting errors, shall we 
1) use `require.NoError(t, err)`
2) register `t.Cleanup` function to stop the server eventually?

Same suggestion applies below (no collecting of errors).